### ISSUE TITLE
Consider supporting alternate content blocks at root of part content

### ISSUE THREAD

**Issue by twsouthwick** [OfficeDev/Open-XML-SDK#1291]
If a part has an `mc:AlternateContent` block at the root of the document, we do not load it. This is loaded fine in Office, so we may want to enable this scenario.

**Comment by tomjebo** [comment]
I've done some digging on this including the impetus for this issue. Some notes here: 

TLDR: I think this is a good idea in general. To add the ability to have any part contain `mc:AlternateContent` as its root element is legal according to ISO 29500-3:2015 7.5 which has a note specifically calling this out: 

> [Note: The AlternateContent element can appear as the root element of a markup document. end note]

Therefore, we should be able to handle this situation without exception or failure.

More information:
1. The initial issue that sparked this investigation ends up being a mistake (i.e. bug, rather probably two bugs) in that PowerPoint inadvertently encapsulated the `sldMaster` root element of the slideMaster1.xml part with an `mc:Choice`, placing the enclosing `mc:AlternateContent` element as the root of the part. It appears it was intending to wrap only the `transition` element with the AC block. This resulted in a badly formed SlideMaster part due to the fact that PowerPoint did not fully provide the Fallback (or other Choices) but provided a Fallback with only the `transition` element and nothing else. It doesn't appear that `mc:AlternateContent` has been used validly in the past for SlideMaster parts and so the SDK code was not written to accommodate this. Hence the failure.

2.  the failure is in OpenXmlPartRootElement.cs, specifically, this code: 

```csharp
if (!resolver.IsKnown(qname.Namespace) || !QName.Equals(qname))
                {
                    var elementQName = new XmlQualifiedName(xmlReader.LocalName, xmlReader.NamespaceURI).ToString();
                    var msg = SR.Format(ExceptionMessages.Fmt_PartRootIsInvalid, elementQName, XmlQualifiedName.ToString());

                    throw new InvalidDataException(msg);
                }
```
After this code, presuming that we accept the `mc:AlternateContent` stuff, we need to load the `sldMaster` content. 

3. Since the SDK understands all the namespaces, is not consuming for functionality (i.e. it's not a graphics application or wordprocessor) and is not like previous versions of PowerPoint or other applications that might only understand one of the choice namespaces, we typically would not make a choice and select only one root element. We would leave the full `mc:AlternateContent` block with all choices and fallback intact.  However, this presents a bit of a problem when loading the `sldMaster` content into the "strong typed DOM tree".  We cannot have two or more versions of the part loaded into the DOM tree and we don't currently have a mechanism for a part to contain multiple versions of the content.

Some options for solving this: 

- if no other options, just default to load the latest versioned choice. Not a great path to take as it hides a version(s) of the part.
- allow an option to the caller of Load to provide the version they wish to load into the DOM. Perhaps a little better.
- provide the internal mechanism whereby a part would appear multi-versioned to consuming code. This seems complicated and would be likely some large effort and/or design. But would be the fullest and most accommodating.
- other options? 

4. If the cost of doing this is prohibitive or more than we want to take on, then another approach could be to evaluate whether this is even used in Word/PowerPoint/Excel and if not, hold off implementing it until such a time that Office commits to produce parts with versioned root elements.

**Comment by twsouthwick** [comment]
I already had thought about moving the part loading to its own feature to encapsulate the logic there. Once that is done, it would be much easier to enable this kind of behavior. 

As for the versions, there OpenSettings has the ability to set a version - we could just use that.

**Comment by tomjebo** [comment]
I'm going to check with Office folks to see if they know of this approach being used in any part across all the apps. If not, then I would vote for a design branch that we could prototype or complete a implementation of versioned root elements but not merge it in. The reason is that we don't have a valid Office document to base our design off. AlternateContent root elements are technically legal but we don't have any examples of this. I would want to test them first. Perhaps the Office teams have an example of it.

**Comment by tomjebo** [comment]
Or we could provide it as an experimental version or something.

**Comment by tomjebo** [comment]
- provide the internal mechanism whereby a part would appear multi-versioned to consuming code. This seems complicated and would be likely some large effort and/or design. But would be the fullest and most accommodating.

And to clarify, I think the above option would be the correct way to go. The SDK should load packages with access to all XML parts and DOM, not just one namespace/version. So this would entail modifying the part objects to include collections of versioned XML.

**Comment by twsouthwick** [comment]
That's the beautify of the "Features" infrastructure - things can be opted in to for experimental stuff. Once #1294 is completed, it will be much easier (currently much of this stuff is handled before we return control to the user)

**Comment by tomjebo** [comment]
So what I'm thinking of is something like an OpenXmlVersionedPart similar to OpenXmlPart which contains a collection of OpenXmlParts for the various versions. From the package and part container perspective it would look like OpenXmlPart but it would specifically have the functionality to navigate the AC versioning. Still thinking about this but just an idea.

**Comment by tomjebo** [comment]
excerpt from internal discussion on the originating bug: 

...if you add a sldMaster child `AlternateContent` block wrapping a `transition` element per [MS-PPTX] 3.1 you can get the XML that we saw in the customer's bug. Depending on whether you call `AppendChild()` or how you call `AddChild()`, it might add that AC block at the end or the beginning or anywhere in the list of children because it does not know what order to place an `AlternateContent` element. Only a `Transition` element and we don’t dig into AC blocks when adding them. **This is also something we could add as rich features to the SDK.**
